front/src/applications/operationalStudies/utils.ts

```diff
@@ -209,3 +214,79 @@ export const convertDepartureTimeIntoSec = (departureTime: string) => {
 
 export const isInvalidName = (name?: string | null) =>
   !name || name.length > SMALL_INPUT_MAX_LENGTH;
+
+/**
+ * Check if the scheduled points are honored with a trainSchedule and a trainSummary
+ * @param trainResult
+ * @param trainSummary
+ * @returns true if the scheduled points are not honored
+ */
+export const isScheduledPointsNotHonored = (
+  trainSchedule: TrainScheduleResult,
+  trainSummary: Extract<SimulationSummaryResult, { status: 'success' }>
+): boolean => {
+  if (trainSummary.path_item_times_final.length !== trainSchedule.path.length) {
+    throw new Error('Asserstion failed');
+  }
+  if (!trainSchedule.schedule) return false;
+  return trainSchedule.schedule.some((schedule, index) => {
+    if (schedule.arrival === null) return false;
+    // need to add +1 because origin will never be in schedules
+    return (
+      Math.abs(
+        sToMs(ISO8601Duration2sec(schedule.arrival as string)) -
+          trainSummary.path_item_times_final[index + 1]
+      ) >= TOLERANCES
+    );
+  });
+};
+
+export const pathItemMapping = (trainResult: TrainScheduleResult) => {
+  const pathItemMap: { [key: string]: number } = {};
+
+  trainResult.path.forEach((pathItem, index) => {
+    if (pathItemMap[pathItem.id]) throw new Error('Assertion failed');
+    pathItemMap[pathItem.id] = index;
+  });
+
+  return pathItemMap;
+};
+
+/**
+ * Check if the train is too fast with a trainSchedule and a trainSummary
+ * @param trainResult
+ * @param trainSummary
+ * @returns true if the train is too fast
+ */
+export const isToofast = (
+  trainSchedule: TrainScheduleResult,
+  trainSummary: Extract<SimulationSummaryResult, { status: 'success' }>
+): boolean => {
+  if (
+    trainSummary.path_item_times_final.length !== trainSummary.path_item_times_provisional.length
+  ) {
+    throw new Error('Assertion failed');
+  }
+
+  const toCheckPathItemIds = new Set<string>(trainSchedule.margins?.boundaries);
+  trainSchedule.schedule?.forEach((schedule) => {
+    if (schedule.arrival !== null || schedule.stop_for !== null) {
+      toCheckPathItemIds.add(schedule.at);
+    }
+  });
+  toCheckPathItemIds.add(trainSchedule.path[trainSchedule.path.length - 1].id);
+
+  if (!toCheckPathItemIds.size) return false;
+
+  const pathItemMap = pathItemMapping(trainSchedule);
+
+  // eslint-disable-next-line no-restricted-syntax
+  for (const pathItemId of toCheckPathItemIds) {
+    const pathItemIndex = pathItemMap[pathItemId];
+    const pathItemTimeFinal = trainSummary.path_item_times_final[pathItemIndex];
+    const pathItemTimeProvisional = trainSummary.path_item_times_provisional[pathItemIndex];
+
+    if (pathItemTimeProvisional > pathItemTimeFinal + TOLERANCES) return true;
+  }
+  return false;
```

Comment:
You can use the function some instead of this loop, like in isScheduledPointsNotHonored ;)
```suggestion
  return [...toCheckPathItemIds].some((pathItemId) => {
    const pathItemIndex = pathItemMap[pathItemId];
    const pathItemTimeFinal = trainSummary.path_item_times_final[pathItemIndex];
    const pathItemTimeProvisional = trainSummary.path_item_times_provisional[pathItemIndex];

    if (pathItemTimeProvisional > pathItemTimeFinal + TOLERANCES) return true;
    return false;
  });
```

Reply:
>  if (pathItemTimeProvisional > pathItemTimeFinal + TOLERANCES) return true;
> return false;

Bad pattern:

```
return (pathItemTimeProvisional > pathItemTimeFinal + TOLERANCES);
```

What is the behavior of using `return` in a lambda?

Reply:
Do we have optimization when using an iterator instead of a for loop?

Nitick: In my opinion the first version is more readable :shrug:

Reply:
> Bad pattern:
> 
> ```
> return (pathItemTimeProvisional > pathItemTimeFinal + TOLERANCES);
> ```
You're right about this; I was considering the for loop and wondering whether .some() would be better or not, and I missed it :/

From what I have read, a for loop is as efficient as .some(). I usually prefer to use the available functions of JavaScript, but @emersion also thinks that it is more readable like this, so we can keep it like this 👍

Reply:
TBH the speed here likely doesn't matter at all, we shouldn't be micro-optimizing :)

Reply:
Done